doc/manual/src/language/import-from-derivation.md

```diff
@@ -0,0 +1,139 @@
+# Import From Derivation
+
+The value of a Nix expression can depend on the contents of a [store object].
+In this case, when that store object is needed, evaluation will be paused, the store object [realised], and then evaluation resumed.
+
+[store object]: @docroot@/glossary.md#gloss-store-object
+[derivation]: @docroot@/glossary.md#gloss-derivation
+[realised]: @docroot@/glossary.md#gloss-realise
+
+This has performance implications:
+Since evaluation is sequential, each required store object that is not already in the store will also be realised sequentially.
+Usually, if store objects are not already present, realisation is orders of magnitude slower than evaluation.
+
+Passing an expression `expr` that evaluates to a [store path](@docroot@/glossary.md#gloss-store-path) to any built-in function which reads from the filesystem constitutes Import From Derivation:
+
+- [`import`](./builtins.md#builtins-import)` expr`
+- [`builtins.readFile`](./builtins.md#builtins-readFile)` expr`
+- [`builtins.readFileType`](./builtins.md#builtins-readFileType)` expr`
+- [`builtins.readDir`](./builtins.md#builtins-readDir)` expr`
+- [`builtins.pathExists`](./builtins.md#builtins-pathExists)` expr`
+- [`builtins.filterSource`](./builtins.md#builtins-filterSource)` f expr`
+- [`builtins.path`](./builtins.md#builtins-path)` { path = expr; }`
+- [`builtins.hashFile`](./builtins.md#builtins-hashFile)` t expr`
+- `builtins.scopedImport x drv`
+
+Realising store objects during evaluation can be disabled by setting [`allow-import-from-derivation`](../command-ref/conf-file.md#conf-allow-import-from-derivation) to `false`.
+Without IFD it is ensured that evaluation is complete and Nix can produce a build plan before starting any realisation.
+
+## Example
+
+In the following Nix expression, the inner derivation `drv` produces a file with contents `hello`.
+
+```nix
+# IFD.nix
+let
+  drv = derivation {
+    name = "hello";
+    builder = "/bin/sh";
+    args = [ "-c" "echo -n hello > $out" ];
+    system = builtins.currentSystem;
+  };
+in "${builtins.readFile drv} world"
+```
+
+```shellSession
+nix-instantiate IFD.nix --eval --read-write-mode
+```
+
+```
+building '/nix/store/348q1cal6sdgfxs8zqi9v8llrsn4kqkq-hello.drv'...
+"hello world"
+```
+
+The contents of the derivation's [output path](@docroot@/glossary.md#gloss-output-path) have to be [realised] before they can be read with [`readFile`](./builtins.md#builtins-readFile).
```

Comment:
```suggestion
The contents of the derivation's [output](@docroot@/glossary.md#gloss-output) have to be [realised] before they can be read with [`readFile`](./builtins.md#builtins-readFile).
```
This is scope creep, but it would be good to change that entry. "output path" is not a great: it is just a *store path* to a *derivation output*; there is no special notion of "output path", it is just the combination of those other two concepts.

Moreover, the contents of the "output *path*" is not `hello world`, it is `asidfaosidjfoiasdfu-hello`! :grin: